internal/wasm/host.go

```diff
@@ -97,57 +97,53 @@ func (s *Store) NewHostModule(moduleName string, nameToGoFunc map[string]interfa
 	}
 
 	exportCount := len(nameToGoFunc)
-
 	ret := &HostModule{name: moduleName, NameToFunctionInstance: make(map[string]*FunctionInstance, exportCount)}
 	hostModule := &ModuleInstance{
 		Name:       moduleName,
 		Exports:    make(map[string]*ExportInstance, exportCount),
 		hostModule: ret,
 	}
-	s.moduleInstances[moduleName] = hostModule
+
 	for name, goFunc := range nameToGoFunc {
-		if hf, err := NewGoFunc(name, goFunc); err != nil {
+		hf, err := NewGoFunc(name, goFunc)
+		if err != nil {
 			return nil, err
-		} else if function, err := s.addHostFunction(hostModule, hf); err != nil {
+		}
+
+		f := &FunctionInstance{
+			Name:           fmt.Sprintf("%s.%s", hostModule.Name, hf.wasmFunctionName),
+			HostFunction:   hf.goFunc,
+			FunctionKind:   hf.functionKind,
+			ModuleInstance: hostModule,
+		}
+		hostModule.Exports[hf.wasmFunctionName] = &ExportInstance{Type: ExternTypeFunc, Function: f}
+		hostModule.Functions = append(hostModule.Functions, f)
+		ret.NameToFunctionInstance[name] = f
+
+		if err = s.compileFunction(f, hf); err != nil {
```

Comment:
what I noticed after this is the only things that need the store (besides the initial check for "all clear" on the module name) is this which compiles the function and at the bottom, adding it to the module.

I'm thinking later we can have a similar relationship for host modules as wasm defined ones in terms of preparation errors vs instantiation ones.

Ex. Wasm-defined modules can fail early on decode for any reason (including compilation if we push it up that early)

There seems to be an option for failing early on host modules, for example if the signatures are invalid. Some similar "compile" prep may be useful in the future vs only being able to tell if something will fail when you use it.